ADR/adr-0008-secret-mgmt-for-user-workloads.md

```diff
@@ -0,0 +1,60 @@
+# Secret Management For User Workloads
+
+Date: 2022-10-28
+
+## Status
+
+Accepted
+
+## Context
+
+When user workloads are deployed to environments, the system should be able to provide a way to inject values that is specific to the environment. Today, this is done through environment variables that are managed as overlays on the GitOps repository for the application. However, this method does not provide a good way to manage secrets. This ADR addresses the secret management of user workloads for different environments.
+
+## Decision
+
+### Architecture Overview
+
+All user workloads are synced through GitOps service to their environments. StoneSoup's GitOps service uses the [Argo CD Vault Plugin (AVP)](https://github.com/argoproj-labs/argocd-vault-plugin) to replace secret values from the supported secret management backends.
```

Comment:
I have a concern with using a plugin to handle secrets management. Unfortunately, plugins (including ACP) are not really suitable for a multi-tenant environment. This is due to the fact that the repository server, where the plugin is executed (either within the repository server's pod itself, or in a sidecar), is not tenant-aware. So basically, any secret that is made available to the repository server, is potentially accessible by all tenants.

This is one of the current Argo CD design limitations. There's a [proposal for Argo CD](https://github.com/argoproj/argo-cd/pull/12123) to change the way how Argo CD handles manifest generation. One of its goals is that multi-tenant systems requiring the repository server to process secret information can securely do so.

However, even with the proposal accepted and the appropriate changes incorporated into the code, we'd some heavy architectural lifting on our side to leverage that.

Reply:
What do you think about the method described in [here](https://argocd-vault-plugin.readthedocs.io/en/stable/config/#multitenancy)?

Reply:
I think that's more of a placebo multi-tenancy setup, to be honest. It _might_ be ok when the teams trust each other to a certain extent, however, we're in the fully untrusted domain here. Also, I was wondering what would prevent anyone from setting the `AVP_SECRET` in their `Application` resource to the name of another secret (considering that people usually have kind of control over their own `Application` resources).

Another factor is, the repository server itself should not have access to Kubernetes API in the first place for security reasons (that is why in the upstream's [default configuration](https://github.com/argoproj/argo-cd/blob/9fb7847f4e054b05b222a6cfabe4cd821d90f6e5/manifests/base/repo-server/argocd-repo-server-deployment.yaml#L19), access to Kubernetes for the repository server is disabled). We should avoid making it a requirement for the repository server to be able to talk to Kubernetes.

Also, the repository server does not have any concept of multi-tenancy, it just executes potentially untrusted processes on behalf of the application controller to render manifests and stream the results back for consumption. Now once we let users bring their own GitOps repositories, we won't have control over the repositories' contents anymore, leading to all kind of attack surfaces (Kustomize reading a file on the repository server or the sidecar, Helm reading a resource from Kubernetes, etc).

IMHO, we should not have our users' crown jewels handled by a third-party plugin. The plugin approach is fine for personal or non-critical purposes, but for large scale production environments with untrusted tenants such as ours, I'd suggest using an externalized secrets processing tool such as [external-secrets](https://github.com/external-secrets/external-secrets).

Reply:
I think we are moving ahead *without* the plugin - exploring new designs now. Closing this.